src/ott/neural/layers.py

```diff
@@ -81,59 +78,82 @@ def __call__(self, inputs: jnp.ndarray) -> jnp.ndarray:
 
 
 class PosDefPotentials(nn.Module):
-  r"""A layer to output :math:`\frac{1}{2} ||A_i^T (x - b_i)||^2_i` potentials.
+  """A layer to output  0.5 x^T(A_i A_i^T + Diag(d_i^2))x + b_i^T x + c_i potentials.
 
   Args:
-    use_bias: whether to add a bias to the output.
-    dtype: the dtype of the computation.
-    precision: numerical precision of computation see `jax.lax.Precision`
-      for details.
-    kernel_init: initializer function for the weight matrix.
-    bias_init: initializer function for the bias.
-  """
-  dim_data: int
+  num_potentials: the dimension of the output
+  rank: the rank of the matrix used for the quadratic layer
+  use_linear: whether to add a linear layer to the output
+  use_bias: whether to add a bias to the output.
+  dtype: the dtype of the computation.
+  precision: numerical precision of computation see `jax.lax.Precision` for details.
+  kernel_quadratic_init: initializer function for the weight matrix of the quadratic layer.
+  kernel_linear_init: initializer function for the weight matrix of the linea layer.
+  bias_init: initializer function for the bias.
+  """  # noqa: E501
+
   num_potentials: int
+  rank: int = 0
+  use_linear: bool = True
   use_bias: bool = True
   dtype: Any = jnp.float32
   precision: Any = None
-  kernel_init: Optional[Callable[[PRNGKey, Shape, Dtype], Array]] = None
-  bias_init: Callable[[PRNGKey, Shape, Dtype], Array] = nn.initializers.zeros
+  kernel_quadratic_init: Callable[[PRNGKey, Shape, Dtype],
+                                  Array] = nn.initializers.lecun_normal()
+  kernel_diagonal_init: Callable[[PRNGKey, Shape, Dtype],
+                                 Array] = nn.initializers.ones
+  kernel_linear_init: Callable[[PRNGKey, Shape, Dtype],
+                               Array] = nn.initializers.lecun_normal()
+  bias_init: Callable[[PRNGKey, Shape, Dtype],
+                      Array] = nn.initializers.lecun_normal()
 
   @nn.compact
   def __call__(self, inputs: jnp.ndarray) -> jnp.ndarray:
     """Apply a few quadratic forms.
 
     Args:
-      inputs: Array to be transformed (possibly batched).
+    inputs: Array to be transformed (possibly batched).
 
     Returns:
-      The transformed input.
+    The transformed input.
     """
-    kernel_init = nn.initializers.lecun_normal(
-    ) if self.kernel_init is None else self.kernel_init
+    dim_data = inputs.shape[-1]
     inputs = jnp.asarray(inputs, self.dtype)
-    kernel = self.param(
-        "kernel", kernel_init,
-        (self.num_potentials, inputs.shape[-1], inputs.shape[-1])
-    )
+    inputs = inputs.reshape((-1, dim_data))
 
-    if self.use_bias:
-      bias = self.param(
-          "bias", self.bias_init, (self.num_potentials, self.dim_data)
+    diag_kernel = self.param(
+        "diag_kernel", self.kernel_diagonal_init,
+        (1, dim_data, self.num_potentials)
+    )
+    # ensures the diag_kernel parameter stays non negative
+    diag_kernel = nn.activation.relu(diag_kernel)
```

Comment:
ok, I agree with this, then we should remove the  `^2` of the `d_i^2` in the pydoc